internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,211 @@
+	}
+	if isUpgrading {
+		log.Info(fmt.Sprintf("Preflight check %q failed: ControlPlane %s is upgrading", clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) kubernetesVersionPreflightCheck(ctx context.Context, cpSemver semver.Version, msSemver semver.Version) ctrl.Result {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Check the Kubernetes version skew policy.
+	// => MS minor version cannot be greater than the Control Plane minor version.
+	// => MS minor version cannot be older than 2 minor versions of Control Plane.
+	// Kubernetes skew policy: https://kubernetes.io/releases/version-skew-policy/#kubelet
+	floorVersion := semver.Version{
+		Major: cpSemver.Major,
+		Minor: cpSemver.Minor - 2,
+		Patch: 0,
+	}
+	if msSemver.GT(cpSemver) || msSemver.LT(floorVersion) {
+		log.Info(fmt.Sprintf("Preflight check %q failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to the kubernetes version skew policy", clusterv1.MachineSetPreflightCheckKubernetesVersion, msSemver.String(), cpSemver.String()))
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}
+	}
+
+	return ctrl.Result{}
+}
+
+func (r *Reconciler) kubeadmVersionPreflightCheck(ctx context.Context, cpSemver semver.Version, msSemver semver.Version, ms *clusterv1.MachineSet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// If the bootstrap.configRef is nil return early.
+	if ms.Spec.Template.Spec.Bootstrap.ConfigRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// If using kubeadm bootstrap provider, check the kubeadm version skew policy.
+	// => MS version should match (major+minor) the Control Plane version.
+	// kubeadm skew policy: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#kubeadm-s-skew-against-kubeadm
+	bootstrapConfigRef := ms.Spec.Template.Spec.Bootstrap.ConfigRef
+	groupVersion, err := schema.ParseGroupVersion(bootstrapConfigRef.APIVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q preflight check: failed to parse APIVersion %s", clusterv1.MachineSetPreflightCheckKubeadmVersion, bootstrapConfigRef.APIVersion)
+	}
+	kubeadmBootstrapProviderUsed := bootstrapConfigRef.Kind == "KubeadmConfigTemplate" &&
+		groupVersion.Group == bootstrapv1.GroupVersion.Group
+	if kubeadmBootstrapProviderUsed {
+		if cpSemver.Major != msSemver.Major || cpSemver.Minor != msSemver.Minor {
+			log.Info(fmt.Sprintf("Preflight check %q failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to kubeadm version skew policy: kubeadm bootstrap provider only supports joining with the same (major+minor) version as the control plane", clusterv1.MachineSetPreflightCheckKubeadmVersion, msSemver.String(), cpSemver.String()))
```

Comment:
```suggestion
			log.Info(fmt.Sprintf("Preflight check %q failed: MachineSet version (%s) and ControlPlane version (%s) do not conform to kubeadm version skew policy: kubeadm only supports joining with the same (major+minor) version as the control plane", clusterv1.MachineSetPreflightCheckKubeadmVersion, msSemver.String(), cpSemver.String()))
```
nit: Only to keep it a bit shorter